File: binding/include/charts/libqpieseries.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qpieseries.html
pub const qpieseries = struct {

    /// New constructs a new QPieSeries object.
    pub fn New() QtC.QPieSeries {
        return qtc.QPieSeries_new();
    }


    /// New2 constructs a new QPieSeries object.
    pub fn New2(parent: ?*anyopaque) QtC.QPieSeries {
        return qtc.QPieSeries_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPieSeries_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpieseries.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Type(self: ?*anyopaque, ) i32 {
        return qtc.QPieSeries_Type(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnType(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QPieSeries_OnType(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseType(self: ?*anyopaque, ) i32 {
        return qtc.QPieSeries_QBaseType(@ptrCast(self));
    }

    pub fn Append(self: ?*anyopaque, slice: ?*anyopaque) bool {
        return qtc.QPieSeries_Append(@ptrCast(self), @ptrCast(slice));
    }

    pub fn Append2(self: ?*anyopaque, slices: []const u8) bool {
        return qtc.QPieSeries_Append2(@ptrCast(self), @ptrCast(slices));
    }

    pub fn OperatorShiftLeft(self: ?*anyopaque, slice: ?*anyopaque) QtC.QPieSeries {
        return qtc.QPieSeries_OperatorShiftLeft(@ptrCast(self), @ptrCast(slice));
    }

    pub fn Append3(self: ?*anyopaque, label: []const u8, value: f64) QtC.QPieSlice {
        const label_str = qtc.libqt_string{
    .len = label.len,
    .data = label.ptr,
};
return qtc.QPieSeries_Append3(@ptrCast(self), label_str, value);
    }

    pub fn Insert(self: ?*anyopaque, index: i32, slice: ?*anyopaque) bool {
        return qtc.QPieSeries_Insert(@ptrCast(self), index, @ptrCast(slice));
    }

    pub fn Remove(self: ?*anyopaque, slice: ?*anyopaque) bool {
        return qtc.QPieSeries_Remove(@ptrCast(self), @ptrCast(slice));
    }

    pub fn Take(self: ?*anyopaque, slice: ?*anyopaque) bool {
        return qtc.QPieSeries_Take(@ptrCast(self), @ptrCast(slice));
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QPieSeries_Clear(@ptrCast(self));
    }

    pub fn Slices(self: ?*anyopaque, ) []const u8 {
        return qtc.QPieSeries_Slices(@ptrCast(self));
    }

    pub fn Count(self: ?*anyopaque, ) i32 {
        return qtc.QPieSeries_Count(@ptrCast(self));
    }

    pub fn IsEmpty(self: ?*anyopaque, ) bool {
        return qtc.QPieSeries_IsEmpty(@ptrCast(self));
    }

    pub fn Sum(self: ?*anyopaque, ) f64 {
        return qtc.QPieSeries_Sum(@ptrCast(self));
    }

    pub fn SetHoleSize(self: ?*anyopaque, holeSize: f64) void {
        qtc.QPieSeries_SetHoleSize(@ptrCast(self), holeSize);
    }

    pub fn HoleSize(self: ?*anyopaque, ) f64 {
        return qtc.QPieSeries_HoleSize(@ptrCast(self));
    }

    pub fn SetHorizontalPosition(self: ?*anyopaque, relativePosition: f64) void {
        qtc.QPieSeries_SetHorizontalPosition(@ptrCast(self), relativePosition);
    }

    pub fn HorizontalPosition(self: ?*anyopaque, ) f64 {
        return qtc.QPieSeries_HorizontalPosition(@ptrCast(self));
    }

    pub fn SetVerticalPosition(self: ?*anyopaque, relativePosition: f64) void {
        qtc.QPieSeries_SetVerticalPosition(@ptrCast(self), relativePosition);
    }

    pub fn VerticalPosition(self: ?*anyopaque, ) f64 {
        return qtc.QPieSeries_VerticalPosition(@ptrCast(self));
    }

    pub fn SetPieSize(self: ?*anyopaque, relativeSize: f64) void {
        qtc.QPieSeries_SetPieSize(@ptrCast(self), relativeSize);
    }

    pub fn PieSize(self: ?*anyopaque, ) f64 {
        return qtc.QPieSeries_PieSize(@ptrCast(self));
    }

    pub fn SetPieStartAngle(self: ?*anyopaque, startAngle: f64) void {
        qtc.QPieSeries_SetPieStartAngle(@ptrCast(self), startAngle);
    }

    pub fn PieStartAngle(self: ?*anyopaque, ) f64 {
        return qtc.QPieSeries_PieStartAngle(@ptrCast(self));
    }

    pub fn SetPieEndAngle(self: ?*anyopaque, endAngle: f64) void {
        qtc.QPieSeries_SetPieEndAngle(@ptrCast(self), endAngle);
    }

    pub fn PieEndAngle(self: ?*anyopaque, ) f64 {
        return qtc.QPieSeries_PieEndAngle(@ptrCast(self));
    }

    pub fn SetLabelsVisible(self: ?*anyopaque, ) void {
        qtc.QPieSeries_SetLabelsVisible(@ptrCast(self));
    }

    pub fn SetLabelsPosition(self: ?*anyopaque, position: i32) void {
        qtc.QPieSeries_SetLabelsPosition(@ptrCast(self), @intCast(position));
    }

    pub fn Added(self: ?*anyopaque, slices: []const u8) void {
        qtc.QPieSeries_Added(@ptrCast(self), @ptrCast(slices));
    }

    pub fn OnAdded(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) void) void {
        qtc.QPieSeries_Connect_Added(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Removed(self: ?*anyopaque, slices: []const u8) void {
        qtc.QPieSeries_Removed(@ptrCast(self), @ptrCast(slices));
    }

    pub fn OnRemoved(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) void) void {
        qtc.QPieSeries_Connect_Removed(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Clicked(self: ?*anyopaque, slice: ?*anyopaque) void {
        qtc.QPieSeries_Clicked(@ptrCast(self), @ptrCast(slice));
    }

    pub fn OnClicked(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QPieSeries_Connect_Clicked(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Hovered(self: ?*anyopaque, slice: ?*anyopaque, state: bool) void {
        qtc.QPieSeries_Hovered(@ptrCast(self), @ptrCast(slice), state);
    }

    pub fn OnHovered(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, bool) callconv(.c) void) void {
        qtc.QPieSeries_Connect_Hovered(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Pressed(self: ?*anyopaque, slice: ?*anyopaque) void {
        qtc.QPieSeries_Pressed(@ptrCast(self), @ptrCast(slice));
    }

    pub fn OnPressed(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QPieSeries_Connect_Pressed(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Released(self: ?*anyopaque, slice: ?*anyopaque) void {
        qtc.QPieSeries_Released(@ptrCast(self), @ptrCast(slice));
    }

    pub fn OnReleased(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QPieSeries_Connect_Released(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn DoubleClicked(self: ?*anyopaque, slice: ?*anyopaque) void {
        qtc.QPieSeries_DoubleClicked(@ptrCast(self), @ptrCast(slice));
    }

    pub fn OnDoubleClicked(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QPieSeries_Connect_DoubleClicked(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn CountChanged(self: ?*anyopaque, ) void {
        qtc.QPieSeries_CountChanged(@ptrCast(self));
    }

    pub fn OnCountChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QPieSeries_Connect_CountChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn SumChanged(self: ?*anyopaque, ) void {
        qtc.QPieSeries_SumChanged(@ptrCast(self));
    }

    pub fn OnSumChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QPieSeries_Connect_SumChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPieSeries_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpieseries.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPieSeries_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpieseries.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetLabelsVisible1(self: ?*anyopaque, visible: bool) void {
        qtc.QPieSeries_SetLabelsVisible1(@ptrCast(self), visible);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPieSeries_Delete(@ptrCast(self));
    }
};

File: binding/src/charts/libqpieseries.cpp
#include <QAbstractSeries>
#include <QObject>
#include <QPieSeries>
#include <QPieSlice>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qpieseries.h>
#include "libqpieseries.h"
#include "libqpieseries.hxx"

QPieSeries* QPieSeries_new() {
	 return new VirtualQPieSeries();
}

QPieSeries* QPieSeries_new2(QObject* parent) {
	 return new VirtualQPieSeries(parent);
}

libqt_string QPieSeries_Tr(const char* s) {
	QString _ret = QPieSeries::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QPieSeries_Type(const QPieSeries* self) {
	return self->type();
}

bool QPieSeries_Append(QPieSeries* self, QPieSlice* slice) {
	return self->append(slice);
}

bool QPieSeries_Append2(QPieSeries* self, const libqt_list slices) {
	return self->append(*slices);
}

QPieSeries* QPieSeries_OperatorShiftLeft(QPieSeries* self, QPieSlice* slice) {
	return new QPieSeries(self->operator<<(slice));
}

QPieSlice* QPieSeries_Append3(QPieSeries* self, const libqt_string label, double value) {
	return self->append(QString::fromUtf8(label.data, label.len), value);
}

bool QPieSeries_Insert(QPieSeries* self, int index, QPieSlice* slice) {
	return self->insert(index, slice);
}

bool QPieSeries_Remove(QPieSeries* self, QPieSlice* slice) {
	return self->remove(slice);
}

bool QPieSeries_Take(QPieSeries* self, QPieSlice* slice) {
	return self->take(slice);
}

void QPieSeries_Clear(QPieSeries* self) {
	self->clear();
}

libqt_list QPieSeries_Slices(const QPieSeries* self) {
	return self->slices();
}

int QPieSeries_Count(const QPieSeries* self) {
	return self->count();
}

bool QPieSeries_IsEmpty(const QPieSeries* self) {
	return self->isEmpty();
}

double QPieSeries_Sum(const QPieSeries* self) {
	return self->sum();
}

void QPieSeries_SetHoleSize(QPieSeries* self, double holeSize) {
	self->setHoleSize(holeSize);
}

double QPieSeries_HoleSize(const QPieSeries* self) {
	return self->holeSize();
}

void QPieSeries_SetHorizontalPosition(QPieSeries* self, double relativePosition) {
	self->setHorizontalPosition(relativePosition);
}

double QPieSeries_HorizontalPosition(const QPieSeries* self) {
	return self->horizontalPosition();
}

void QPieSeries_SetVerticalPosition(QPieSeries* self, double relativePosition) {
	self->setVerticalPosition(relativePosition);
}

double QPieSeries_VerticalPosition(const QPieSeries* self) {
	return self->verticalPosition();
}

void QPieSeries_SetPieSize(QPieSeries* self, double relativeSize) {
	self->setPieSize(relativeSize);
}

double QPieSeries_PieSize(const QPieSeries* self) {
	return self->pieSize();
}

void QPieSeries_SetPieStartAngle(QPieSeries* self, double startAngle) {
	self->setPieStartAngle(startAngle);
}

double QPieSeries_PieStartAngle(const QPieSeries* self) {
	return self->pieStartAngle();
}

void QPieSeries_SetPieEndAngle(QPieSeries* self, double endAngle) {
	self->setPieEndAngle(endAngle);
}

double QPieSeries_PieEndAngle(const QPieSeries* self) {
	return self->pieEndAngle();
}

void QPieSeries_SetLabelsVisible(QPieSeries* self) {
	self->setLabelsVisible();
}

void QPieSeries_SetLabelsPosition(QPieSeries* self, int position) {
	self->setLabelsPosition(static_cast<QPieSlice::LabelPosition>(position));
}

void QPieSeries_Added(QPieSeries* self, const libqt_list slices) {
	self->added(*slices);
}

void QPieSeries_Connect_Added(QPieSeries* self, intptr_t slot) {
    void (*slotFunc)(QPieSeries*, libqt_list) = reinterpret_cast<void (*)(QPieSeries*, libqt_list)>(slot);
    QPieSeries::connect(self, &QPieSeries::added, [self, slotFunc](const QList<QPieSlice *>& slices) {
	slotFunc(self, slices);
    });
}

void QPieSeries_Removed(QPieSeries* self, const libqt_list slices) {
	self->removed(*slices);
}

void QPieSeries_Connect_Removed(QPieSeries* self, intptr_t slot) {
    void (*slotFunc)(QPieSeries*, libqt_list) = reinterpret_cast<void (*)(QPieSeries*, libqt_list)>(slot);
    QPieSeries::connect(self, &QPieSeries::removed, [self, slotFunc](const QList<QPieSlice *>& slices) {
	slotFunc(self, slices);
    });
}

void QPieSeries_Clicked(QPieSeries* self, QPieSlice* slice) {
	self->clicked(slice);
}

void QPieSeries_Connect_Clicked(QPieSeries* self, intptr_t slot) {
    void (*slotFunc)(QPieSeries*, QPieSlice*) = reinterpret_cast<void (*)(QPieSeries*, QPieSlice*)>(slot);
    QPieSeries::connect(self, &QPieSeries::clicked, [self, slotFunc](QPieSlice* slice) {
	slotFunc(self, slice);
    });
}

void QPieSeries_Hovered(QPieSeries* self, QPieSlice* slice, bool state) {
	self->hovered(slice, state);
}

void QPieSeries_Connect_Hovered(QPieSeries* self, intptr_t slot) {
    void (*slotFunc)(QPieSeries*, QPieSlice*, bool) = reinterpret_cast<void (*)(QPieSeries*, QPieSlice*, bool)>(slot);
    QPieSeries::connect(self, &QPieSeries::hovered, [self, slotFunc](QPieSlice* slice, bool state) {
	slotFunc(self, slice, state);
    });
}

void QPieSeries_Pressed(QPieSeries* self, QPieSlice* slice) {
	self->pressed(slice);
}

void QPieSeries_Connect_Pressed(QPieSeries* self, intptr_t slot) {
    void (*slotFunc)(QPieSeries*, QPieSlice*) = reinterpret_cast<void (*)(QPieSeries*, QPieSlice*)>(slot);
    QPieSeries::connect(self, &QPieSeries::pressed, [self, slotFunc](QPieSlice* slice) {
	slotFunc(self, slice);
    });
}

void QPieSeries_Released(QPieSeries* self, QPieSlice* slice) {
	self->released(slice);
}

void QPieSeries_Connect_Released(QPieSeries* self, intptr_t slot) {
    void (*slotFunc)(QPieSeries*, QPieSlice*) = reinterpret_cast<void (*)(QPieSeries*, QPieSlice*)>(slot);
    QPieSeries::connect(self, &QPieSeries::released, [self, slotFunc](QPieSlice* slice) {
	slotFunc(self, slice);
    });
}

void QPieSeries_DoubleClicked(QPieSeries* self, QPieSlice* slice) {
	self->doubleClicked(slice);
}

void QPieSeries_Connect_DoubleClicked(QPieSeries* self, intptr_t slot) {
    void (*slotFunc)(QPieSeries*, QPieSlice*) = reinterpret_cast<void (*)(QPieSeries*, QPieSlice*)>(slot);
    QPieSeries::connect(self, &QPieSeries::doubleClicked, [self, slotFunc](QPieSlice* slice) {
	slotFunc(self, slice);
    });
}

void QPieSeries_CountChanged(QPieSeries* self) {
	self->countChanged();
}

void QPieSeries_Connect_CountChanged(QPieSeries* self, intptr_t slot) {
    void (*slotFunc)(QPieSeries*) = reinterpret_cast<void (*)(QPieSeries*)>(slot);
    QPieSeries::connect(self, &QPieSeries::countChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QPieSeries_SumChanged(QPieSeries* self) {
	self->sumChanged();
}

void QPieSeries_Connect_SumChanged(QPieSeries* self, intptr_t slot) {
    void (*slotFunc)(QPieSeries*) = reinterpret_cast<void (*)(QPieSeries*)>(slot);
    QPieSeries::connect(self, &QPieSeries::sumChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

libqt_string QPieSeries_Tr2(const char* s, const char* c) {
	QString _ret = QPieSeries::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QPieSeries_Tr3(const char* s, const char* c, int n) {
	QString _ret = QPieSeries::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPieSeries_SetLabelsVisible1(QPieSeries* self, bool visible) {
	self->setLabelsVisible(visible);
}

// Base class handler implementation
int QPieSeries_QBaseType(const QPieSeries* self) {
	auto* vqpieseries = dynamic_cast<const VirtualQPieSeries*>(self);
	if (vqpieseries && vqpieseries->isVirtualQPieSeries) {
vqpieseries->setQPieSeries_Type_IsBase(true);
	return vqpieseries->type();
}
}

// Auxiliary method to allow providing re-implementation
void QPieSeries_OnType(const QPieSeries* self, intptr_t slot) {
	auto* vqpieseries = dynamic_cast<const VirtualQPieSeries*>(self);
	if (vqpieseries && vqpieseries->isVirtualQPieSeries) {
vqpieseries->setQPieSeries_Type_Callback(reinterpret_cast<VirtualQPieSeries::QPieSeries_Type_Callback>(slot));
}
}

void QPieSeries_Delete(QPieSeries* self) {
    delete self;
}

